openquake/risklib/riskinput.py

```diff
@@ -164,16 +164,23 @@ def cache_epsilons(dstore, oq, assetcol, crmodel, E):
     if oq.ignore_covs or not crmodel.covs or 'LN' not in crmodel.distributions:
         return
     A = len(assetcol)
+    eps = numpy.zeros((A, E), F32)
     logging.info('Storing the epsilon matrix in %s', dstore.tempname)
     if oq.asset_correlation == '1':
-        numpy.random.seed(oq.master_seed)
-        eps = numpy.array([numpy.random.normal(size=E)] * A)
-    else:
-        seeds = oq.master_seed + numpy.arange(E)
-        eps = numpy.zeros((A, E), F32)
-        for i, seed in enumerate(seeds):
-            numpy.random.seed(seed)
-            eps[:, i] = numpy.random.normal(size=A)
+        # assets of the same taxonomy produce the same epsilons
+        n_assets = numpy.bincount(assetcol['taxonomy'])[1:]
+        assert n_assets.sum() == A
+        seeds = numpy.random.SeedSequence(oq.master_seed).spawn(len(n_assets))
+        start, stop = 0, 0
+        for t, seed in enumerate(seeds):
+            rng = numpy.random.default_rng(seed)
+            stop += n_assets[t]
+            eps[start:stop] = [rng.normal(size=E)] * n_assets[t]
+            start += n_assets[t]
+    else:  # uncorrelated epsilons
```

Comment:
won't the following work just the same for uncorrelated epsilons instead of iterating through each asset?
```python
eps = numpy.random.default_rng(oq.master_seed).normal(size=(A, E))
```